From: Stewart Hildebrand <stewart.hildebrand@dornerworks.com>
Date: Mon, 15 Jul 2019 22:49:20 -0400
Subject: [PATCH 1/2] Add RPi4 (bcm2711) device tree for Xen

Boot memory layout:
0x00200000 - 0x003fffff xen (max size 2MiB)
0x00480000 - 0x013fffff Image (max size 15.5MiB)

uart1, spi1, and spi2 share an IRQ. For debug purposes, it is helpful to assign
uart1 (the mini/aux uart) to Xen, but Xen doesn't handle spi1 and spi2 and thus
tries to pass those to dom0. This results in an error since the shared IRQ was
already assigned to Xen. Disable spi1 and spi2 to allow use of uart1 in Xen.

Add reg-shift and reg-io-width properties to uart1

Unfortunately the pcie driver seems to be broken, so disable it for now.
---
 arch/arm64/boot/dts/broadcom/Makefile         |  1 +
 .../boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts | 32 +++++++++++++++++++
 2 files changed, 33 insertions(+)
 create mode 100644 arch/arm64/boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts

diff --git a/arch/arm64/boot/dts/broadcom/Makefile b/arch/arm64/boot/dts/broadcom/Makefile
index 33c8e599c1d7..42e6a94502cd 100644
--- a/arch/arm64/boot/dts/broadcom/Makefile
+++ b/arch/arm64/boot/dts/broadcom/Makefile
@@ -4,6 +4,7 @@ dtb-$(CONFIG_ARCH_BCM2835) += bcm2837-rpi-3-b.dtb \
 dtb-$(CONFIG_ARCH_BCM2709) += bcm2710-rpi-3-b.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-3-b.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2711-rpi-4-b.dtb
+dtb-$(CONFIG_ARCH_BCM2835) += bcm2711-rpi-4-b-xen.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-3-b-plus.dtb
 dtb-$(CONFIG_ARCH_BCM2709) += bcm2710-rpi-cm3.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-cm3.dtb
diff --git a/arch/arm64/boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts b/arch/arm64/boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts
new file mode 100644
index 000000000000..ec17c5595e49
--- /dev/null
+++ b/arch/arm64/boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts
@@ -0,0 +1,32 @@
+#include "bcm2711-rpi-4-b.dts"
+
+/ {
+	chosen {
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		xen,xen-bootargs = "console=dtuart dtuart=/soc/serial@7e215040 sync_console dom0_mem=512M bootscrub=0";
+
+		dom0 {
+			compatible = "xen,linux-zimage", "xen,multiboot-module";
+			reg = <0x00480000 0x00f80000>;
+		};
+	};
+
+	aliases {
+		/delete-property/ spi1;
+		/delete-property/ spi2;
+	};
+};
+
+/delete-node/ &spi1;
+/delete-node/ &spi2;
+
+&uart1 {
+	reg-shift = <2>;
+	reg-io-width = <4>;
+};
+
+&pcie_0 {
+	status = "disabled";
+};
-- 
2.22.0

